src/sparseml/pytorch/sparsification/pruning/mask_params.py

```diff
@@ -373,9 +373,7 @@ def update_param_masks(self, target: Union[float, List[float]]) -> List[Tensor]:
         )
 
         if self._scorer:
-            self._scorer.update_last_applied_sparsity(
-                tensor_sparsity(self.params_data[0])
-            )
+            self._scorer.update_last_applied_sparsity(tensor_list_sparsity(masks))
 
         return self.set_param_masks(masks)
```

Comment:
Updated to use parameters instead of masks
```suggestion
        mask_diff = self.set_param_masks(masks)

        if self._scorer:
            self._scorer.update_last_applied_sparsity(
                tensor_list_sparsity(self.params_data)
            )

        return mask_diff
```

Reply:
@KSGulin the true sparsity is the sparsity that the pruner imposed on masks and this is what is measured/compared. Making use of weights sparsity is not reliable because of the two following use-cases (we had issues with both of them before and unfortunately learned this the hard way):
1. optimizer can set some weights to zero (this actually happened a lot in experiments with Yolo models) and the sparsity will increase a bit, but since masks are not zero, those weights can become non-zero in the following step and the reported sparsity will decrease again. This creates a super weird behaviour and it doesn't work well with sparsity schedulers which rely on the fact that sparsity is a monotonically increasing function.
2. structured pruners (for example 4-block pruners) rely on the fact that the number of non-zero weights is always divisible by the block-size (in this case 4), and then the weights that are set to zero by optimizer will ruin that structure (i.e. block-pruning won't work anymore)

Reply:
Using the true sparsity from params would definitely be the good idea but in the final stage,  after the pruning run has finished and the model is being exported for deployment (and when final sparsities are being reported). At that point we don't have to care about any structure in the masks as the model will be used just for inference.

Reply:
great points @eldarkurtic, especially for (2) we should be respecting any mask structure. `tensor_list_sparsity` looks like a clean solution